### ISSUE TITLE
Add `no-invalid-position-declaration`

### ISSUE THREAD

**Issue by tim-janik** [stylelint/stylelint#8524]
### What is the problem you're trying to solve?

During some SCSS -> CSS variable porting, I was left with a bunch of CSS variables declared outside :root{} (or any other rule).
I failed to find a stylelint check that catches this.

### What solution would you like to see?

I would like stylelint to provide a check similar to the following:

```js
module.exports = {
  plugins: [
    { ruleName: 'my/no-standalone-custom-properties',
      rule: function (primaryOption, secondaryOptions) {
        return function (root, result) {
          root.walkDecls (decl => {
            if (decl.prop && decl.prop.startsWith ('--') && // CSS var
                ! ['rule', 'atrule'].includes (decl.parent.type)) // outside CSS rule
              {
                if (0) console.error (`CSS custom property must be declared inside a selector: ${decl.prop}\ndecl.parent.type: ${JSON.stringify (decl.parent.type)}$
                result.warn (`CSS custom property must be declared inside a selector: ${decl.prop}`, {
                  node: decl,
                  message: `Encountered standalone custom property: ${decl.prop}: ${decl.value}`
                });
              }
          });
        };
      }
    },
  ],
  rules: {
    'my/no-standalone-custom-properties': true,
  }
};
```

I don't know any stylelint internals so the actual check would probably need more sophistication...

Please let me know if there is any existing configuration that can already catch this.

**Comment by jeddy3** [comment]
@tim-janik Thanks for the request and for using the template.

I'm surprised this request hasn't come up before, as it seems like people would encounter it more often.

Stylesheets are at-rules, qualified rules, comments and whitespace ([spec](https://drafts.csswg.org/css-syntax/#stylesheet-diagram)).

We can add a `declaration-no-invalid-position` rule, similar to our existing [`no-invalid-position-at-import-rule`](https://stylelint.io/user-guide/rules/no-invalid-position-at-import-rule) one. As in your example, it'd disallow declarations outside of rules and at-rules:

```css
--custom-prop: value;
color: red;
```

(In another world, we'd could've [caught this in the parser](https://github.com/csstree/csstree/blob/master/docs/parsing.md#context) but we use a permissive parser for the general structure of the stylesheet and only reach for stricter and more granular parsers for values and preludes so that we can support CSS language extensions like SCSS.)

If there are no objections to the `declaration-no-invalid-position` proposal, I'll label this up.

**Comment by jeddy3** [comment]
As there are no objections, I'll label this up.

- Name: `no-invalid-position-declaration`
- Description: Disallow invalid position declarations.
- Message: `rejected: 'Unexpected invalid position declaration'`
- Primary option: true
- Secondary options: None
- Category: Avoid errors > Invalid
- Autofix?: No
- Included in sharable configs?: Yes, recommended
- Expanded description: Declarations must be nested inside rules and at-rules.

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

You can use the existing `no-invalid-position-at-import-rule` rule as a blueprint.

**Comment by sw1tch3roo** [comment]
Hi! I’d like to work on this issue, let me know if that’s okay =) @jeddy3
